Allow connecting nested collections to multiple=true inputs in workflow editor

mvdbeek opened the pull request:
<img width="1115" alt="screen shot 2019-02-28 at 22 05 30" src="https://user-images.githubusercontent.com/6804901/53599156-cb201c80-3ba6-11e9-8acc-1a0834b2d312.png">

mvdbeek committed:
Allow connecting nested collections to multiple=true inputs in workflow editor

innovate-invent commented:
Should I go ahead and patch my Galaxy instance with this or is it still WIP?

mvdbeek commented:
Sure, give it a try. I haven't tested this yet beyond being able to create the workflow in the editor.

innovate-invent commented:
Just rigging up my workflow now, it seems to delete the list:list connection every time I change a tool parameter

innovate-invent commented:
Shoot, this change allows the connection but Galaxy seems to flatten the list before passing it to the tool.

mvdbeek commented:
We fixed that recently, and I think we also backported that to 18.09

mvdbeek commented:
Nope, just 19.01.

innovate-invent commented:
Shoot, ok, looks like I am upgrading...

mvdbeek commented:
> it seems to delete the list:list connection every time I change a tool parameter

It should only delete the connection if the list type changes, so for instance if you produce `list` or `list:list:list` instead of `list:list`. Simple parameter changes don't change the downstream connection for me. I've played around with this some more and I think it works fine, so taking it out of WIP.

mvdbeek commented:
damn, no travis here ...

innovate-invent commented:
Sorry, to be clear, the tool receiving the list:list breaks its input connection when you change its parameters. This was tested on 18.09 though, I am trying to set up a 19.01 instance.

mvdbeek commented:
What tool is that, and what parameter change causes this?

innovate-invent commented:
iuc progressiveMauve, any of the toggles.

innovate-invent commented:
Just loaded a 19.01 instance with the patch and I get this:
![image](https://user-images.githubusercontent.com/1768820/53663281-3820ca00-3c1a-11e9-931b-bd7233fdfe09.png)

It isnt a draw artifact, moving the outputting tool will update the ghost line to follow it.
Creating the connection causes my browser to freeze up for a second when I release the mouse button. Saving and reloading with this artifact also locks up the browser, its weird though, the webpage is unresponsive except for when hovering over the used space info in the top left. It still pops up the tool tip describing it.

Changing the parameters of the tool doesn't break the connection though.

mvdbeek commented:
Yeah, probably a javascript error, can you open the browser console and see if there are any errors ?

innovate-invent commented:
```
TypeError: this.terminalMapping is undefined[Learn More] libs.chunk.js:182:7 
	setMapOver workflow-terminals.js:171 connect workflow-terminals.js:403 connect workflow-connector.js:23 from_simple workflow-manager.js:266 each jQuery from_simple workflow-manager.js:263 each jQuery from_simple workflow-manager.js:258 each jQuery from_simple workflow-manager.js:256 success workflow-view.js:333 fire4jQuery
```

mvdbeek commented:
I do see the disconnect as well, and it shouldn't happen AFAICT, but the error you posted I can't reproduce, any chance something went wrong building the client, or the browser still using an old version of this ? The changes in workflow-view-terminals.js should prevent this error from occurring.

innovate-invent commented:
Here is the console output while adding and attaching the list:list between the tools.
```
Polyfills are running polyfills.js:28
JQMIGRATE: Migrate is installed with logging active, version 1.4.1 jquery-migrate.js:26
setGalaxyInstance /workflow/editor singleton.js:10
You are running Vue in development mode.
Make sure to turn on production mode when deploying for production.
See more tips at https://vuejs.org/guide/deployment.html vue.runtime.esm.js:8021
setGalaxyInstance /workflow/editor singleton.js:10
unreachable code after return statement[Learn More] libs.chunk.js:142:4
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/ libs.chunk.js:9064:4
Source map error: request failed with status 404
Resource URL: http://vm:9080/static/style/base.css?v=1551459098
Source Map URL: base.css.map[Learn More]
Warning: No collection type or collection type source defined. workflow-terminals.js:555
JQMIGRATE: Can't change the 'type' of an input or button in IE 6/7/8 jquery-migrate.js:48
console.trace() jquery-migrate.js:50

    migrateWarn4jQuery
    render ui-misc.js:94 triggerEvents5Backbone
    render form-input.js:129 initialize form-input.js:49 View2Backbone
    _addRow form-section.js:171 add form-section.js:44 render form-section.js:24 forEach Underscore render form-section.js:23 initialize form-section.js:16 View2Backbone
    _renderForm form-view.js:132 render form-view.js:102 _render tool-form-base.js:89 _update tool-form-base.js:48 then8jQuery
    _update tool-form-base.js:47 execute deferred.js:44 fire3jQuery
    execute deferred.js:41 _update tool-form-base.js:45 initialize tool-form-base.js:25 View3Backbone
    initialize workflow-forms.js:44 View2Backbone
    showForm workflow-view.js:760 node_changed workflow-manager.js:327 init_field_data workflow-node.js:230 success workflow-view.js:648 request utils.js:192 fire4jQuery

Warning: No collection type or collection type source defined. workflow-terminals.js:568
TypeError: this.terminalMapping is undefined[Learn More] libs.chunk.js:182:7 
```

mvdbeek committed:
Only decrement output termninal map over

Otherwise collection connections into multiple="true" inputs
would be destroyed by the `destroyIfInvalid` check.

mvdbeek commented:
I fixed the issue with outputs getting disconnected, but I don't see the error you're seeing.
![connect_and_change_values](https://user-images.githubusercontent.com/6804901/53667435-477f3200-3c71-11e9-8c12-f710f907e3f6.gif)

mvdbeek commented:
Also I'm not sure all of this is going to work on 19.01, there were some additional changes that might be missing if you just add the changes here.

innovate-invent commented:
So I will have to move to dev?

mvdbeek commented:
Maybe try this one again with the new commit added. Also make sure you really got the change in line 80 of `client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js`.

innovate-invent commented:
I added your most recent commit, rebuilt, and everything worked perfectly. Thanks so much for fixing this.

mvdbeek committed:
Don't allow connection to paired input, fix qunit test

mvdbeek committed:
Fix qunit test

jmchilton commented on client/galaxy/scripts/mvc/workflow/workflow-terminals.js:
```diff
@@ -157,14 +157,21 @@ var Terminal = Backbone.Model.extend({
+                // No mapping over necessary
+                return;
+            }
+            output_val = val.effectiveMapOver ? val.effectiveMapOver(description): val;
```
Nifty.
